20220714-lookup-api.md

```diff
@@ -0,0 +1,463 @@
+
+- **What use cases does it support?**
+  Search with Permissions where the number of objects of a particular type the user could have access to is low (~ thousands), and the percentage of the total objects that the user can have access to is also low.
+
+- **What is the expected outcome?**
+  OpenFGA clients use the ListObjects API to return a list of object ids and then the client uses the list to filter and order their object list in their applications (pagination on the filtered list is the clients responsibility). The expectation is that the client consumes the whole ListObjects response before filtering client side, and does not try to partially filter based on a subset of the response from ListObjects. Hence, it is not the goal of ListObjects to be able to continue the ListObjects process from a prior invocation.
+
+# What it is
+[what-it-is]: #what-it-is
+
+The ListObjects API will provide clients with the ability to query the object ids that a user has a particular relationship with. The list of object ids the ListObjects API returns can then be used by a client to filter an object collection to only the objects that a user has a particular relationship with by taking the intersection of it with any client provided filters/ordering of the underlying dataset. This is described in more detail with [Search with Permission (Option 3)](https://openfga.dev/docs/interacting/search-with-permissions#option-3-build-a-list-of-ids-then-search) document.
+
+It will be provided as two different API endpoints: one will provide a streaming endpoint and the other will implement the endpoint without streaming. These two different endpoints could be used interchangably, but clients can choose which one may be better for their integration environment.
+
+Consider a developer/client integrating with OpenFGA with the following authorization model:
+```
+type folder
+    relations
+        define viewer as self
+
+type document
+    relations
+        define viewer as self or editor or viewer from parent
+        define editor as self
+        define parent as self
+```
+and the following relationship tuples:
+
+| object         | relation | user           |
+|----------------|----------|----------------|
+| document:doc1  | viewer   | bob            |
+| document:doc2  | editor   | bob            |
+| document:doc3  | parent   | folder:folder1 |
+| folder:folder1 | viewer   | bob            |
+
+## StreamedListObjects
+StreamedListObjects will implement a grpc [server streaming RPC](https://grpc.io/docs/what-is-grpc/core-concepts/#server-streaming-rpc) endpoint.
+
+The endpoint will stream each object id the target user has the given relationship with as it is evaluated, and the server will continuing streaming results until either the list has been exhausted or the `listObjects-deadline` has been reached (see [Configuration Changes](#configuration-changes)).
```

Comment:
> The endpoint will stream each object id the target user has the given relationship with as it is evaluated, and the server will continuing streaming results until either the list **has been exhausted**

It says below that ListObjectsMaxResults applies to this endpoint as well

Reply:
I've reworded it to include the behavior of the `listObjects-max-results` flag as well. How does that look?